README.md

```diff
@@ -201,7 +201,88 @@ class MyOtherServiceTemplate : EndpointTemplate
+
+- Retries
+  - Retries should be reduced or even disabled for tests.
+  - Otherwise (with the default retry configuration) the IntegrationTests throw a "Some failed messages were not handled by the recoverability feature."-Exception because of a fixed 30 sec timeout in the `NServiceBus.AcceptanceTests.ScenarioRunner`.
+- Cleaning up the queues via `PurgeOnStartup(true)`
+
+#### Generic host support
+
+NServiceBus endpoints can be [hosted using the generic host](https://docs.particular.net/samples/hosting/generic-host/). When using the generic host the endpoint lifecycle and configuration are controlled by the host. The following is a sample endpoint hosted using the generic host:
+
+<!-- snippet: basic-generic-host-endpoint -->
+<a id='snippet-basic-generic-host-endpoint'></a>
+```cs
+public static void Main(string[] args)
+{
+    CreateHostBuilder(args).Build().Run();
+}
+
+public static IHostBuilder CreateHostBuilder(string[] args)
+{
+    var builder = Host.CreateDefaultBuilder(args);
+    builder.UseConsoleLifetime();
+
+    builder.UseNServiceBus(ctx =>
+    {
+        var config = new EndpointConfiguration("endpoint-name");
+        config.UseTransport<LearningTransport>();
+
+        return config;
+    });
+
+    return builder;
+}
+```
+<sup><a href='/src/Snippets/GenericHostSnippets.cs#L27-L48' title='Snippet source file'>snippet source</a> | <a href='#snippet-basic-generic-host-endpoint' title='Start of snippet'>anchor</a></sup>
+<!-- endSnippet -->
+
+> For more information about hosting NServiceBus using the generic host refer to the [official documentation](https://docs.particular.net/samples/hosting/generic-host/). 
+
+Before using generic host hosted endpoints with `NServiceBus.IntegrationTesting`, a minor change to the above snippet is required:
+
+<!-- snippet: basic-generic-host-endpoint-with-config-previewer -->
+<a id='snippet-basic-generic-host-endpoint-with-config-previewer'></a>
+```cs
+public static IHostBuilder CreateHostBuilder(string[] args, Action<EndpointConfiguration> configPreview)
+{
+    var builder = Host.CreateDefaultBuilder(args);
+    builder.UseConsoleLifetime();
+
+    builder.UseNServiceBus(ctx =>
+    {
+        var config = new EndpointConfiguration("endpoint-name");
+        config.UseTransport<LearningTransport>();
+
+        configPreview?.Invoke(config);
+        
+        return config;
+    });
+
+    return builder;
+}
+```
+<sup><a href='/src/Snippets/GenericHostSnippets.cs#L50-L68' title='Snippet source file'>snippet source</a> | <a href='#snippet-basic-generic-host-endpoint-with-config-previewer' title='Start of snippet'>anchor</a></sup>
+<!-- endSnippet -->
+
+The testing engine needs to access the endpoint configuration before it's initialized to register the needed tests behaviors. The creation of the `IHostBuilder` needs to be tweaked to invoke a callback delegate that the test engine injects at tests runtime. 
```

Comment:
Is this something that could be done in `OnGetConfiguration`, too? I used `OnGetConfiguration` to do my test specific configuration adjustments (retries). If yes the production code doesn't have to contain this test specific extension point `configPreview?.Invoke(config);` which might be confusing.

Reply:
> Is this something that could be done in `OnGetConfiguration`, too? I used `OnGetConfiguration` to do my test specific configuration adjustments (retries). If yes the production code doesn't have to contain this test specific extension point `configPreview?.Invoke(config);` which might be confusing.

The extension point is weird, but I couldn't find any other workaround. When using the generic host the lifecycle of the endpoint works more or less in the following way:

1. User code creates the endpoint configuration in `UseNserviceBus(...)`
2. the configuration is used to create a startable endpoint
   - the configuration is frozen
   - the underlying IoC container is configured and frozen
3. the host starts and so the endpoint

When the testing engine gest control back it's too late to manipulate the configuration, without mentioning that it's not stored anywhere accessible so even if it was not too late the configuration is not available. The test engine needs to add behaviors that require changing the configuration and the container, but both are read-only at this point.

The `OnGetConfiguration` doesn't work either because it expects the lifecycle to be controlled by the test engine and not be the generic host. Even if I decide to stop depending on the acceptance testing framework, which sooner or later I will, I don't see anyway a different way of doing it.

I can anyway merge this one, and release 1.1 to get you unstuck and then brainstorm more to see if there are other options.

Reply:
@mauroservienti OK, thanks for the explanation.